src/frontend/src/flows/register/index.ts

```diff
@@ -98,24 +109,82 @@ export const registerFlow = async ({
   const flowStart = precomputeFirst(() => identityRegistrationStart());
 
   const displayUserNumber = displayUserNumberWarmup();
-  const identity = await savePasskeyOrPin();
-  if (identity === undefined) {
-    // TODO: Return something meaningful if getting the identity fails
+  const savePasskeyResult = await savePasskeyOrPin({
+    pinAllowed: await pinAllowed(),
+  });
+  if (savePasskeyResult === "canceled") {
     return "canceled";
   }
-  const alias = await inferPasskeyAlias({
-    authenticatorType: identity.getAuthenticatorAttachment(),
-    userAgent: navigator.userAgent,
-    uaParser,
-  });
+  const result_ = await (async () => {
+    if (savePasskeyResult === "pin") {
+      const pinResult = await setPinFlow();
+      if (pinResult.tag === "canceled") {
+        return "canceled";
+      }
 
-  const authnMethodData = passkeyAuthnMethodData({
-    alias,
-    pubKey: identity.getPublicKey().toDer(),
-    credentialId: identity.rawId,
-    authenticatorAttachment: identity.getAuthenticatorAttachment(),
-  });
-  const authnMethod = "passkey" as const;
+      pinResult.tag satisfies "ok";
+
+      // XXX: this withLoader could be replaced with one that indicates what's happening (like the
+      // "Hang tight, ..." spinner)
+      const { identity, pinIdentityMaterial } = await withLoader(() =>
+        constructPinIdentity(pinResult)
+      );
+      const alias = await inferPinAlias({
+        userAgent: navigator.userAgent,
+        uaParser,
+      });
+      return {
+        identity,
+        authnMethodData: pinAuthnMethodData({
+          alias,
+          pubKey: identity.getPublicKey().toDer(),
+        }),
+        finalizeIdentity: (userNumber: bigint) =>
+          storePinIdentity({ userNumber, pinIdentityMaterial }),
+        finishSlot: tempKeyWarningBox({ i18n: new I18n() }),
+        authnMethod: "pin" as const,
+      };
+    } else {
+      const identity = savePasskeyResult;
+      // TODO: Return something meaningful if getting the passkey identity fails
```

Comment:
This is the only new line that was not just reverted. Because the view can return `undefined`.

This is in the current code as well. The way this error was handled before was to never resolve the promise and show a toast error.